text/0000-heterogeneous-comparisons.md

```diff
@@ -0,0 +1,52 @@
+- Feature Name: heterogeneous_comparisons
+- Start Date: 2017-06-06
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow to compare integer values of different signedness and size.
+
+# Motivation
+[motivation]: #motivation
+
+Right now every comparison between different integral types requires a cast. These casts don't only clutter the code, but also encourage writing incorrect code.
+
+The easiest way to compare signed and unsigned values is to cast unsigned value to signed type. It works most of the time, but it will silently ignore overflows.
+
+Comparison between values of different integer types is always well-defined. There is only one correct result and only one way to get it. Allowing compiler to perform these comparisons will reduce both code clutter and count of hidden overflow/underflow errors users make.
+
+# Detailed design
+[design]: #detailed-design
+
+`Ord` and `Eq` traits should be modified to allow `Rhs` type not equal to `Self`.
+
+After that `PartialEq`, `Eq`, `PartialOrd` and `Ord` should be implemented for all pairs of signed/unsigend 8/16/32/64/(128) bit integers and `isize`/`usize` variants.
+
+Implementation for signed-singed and unsigned-unsigned pairs should promote values to larger type first, then perform comparison.
+
+Implementation for signed-unsigned pairs should first check if signed value less than zero. If not, then it should promote both values to signed type with the same size as larger argument type and perform comparison.
```

Comment:
> I personally find check for non-negativeness a bit cleaner.

It's also less prone to the typo of checking whether a `u32` is greater than `std::u32::MAX` rather than `std::i32::MAX`.